lib/internal/streams/pipeline.js

```diff
@@ -73,25 +72,97 @@ function pipeline(...streams) {
   }
 
   let error;
-  const destroys = streams.map(function(stream, i) {
+  const destroys = [];
+
+  function finish(err, val, final) {
+    if (!error && err) {
+      error = err;
+    }
+
+    if (error || final) {
+      for (const destroy of destroys) {
+        destroy(error);
+      }
+    }
+
+    if (final) {
+      callback(error, val);
+    }
+  }
+
+  async function pump(iterableOrPromise, writable) {
+    try {
+      for await (const chunk of await iterableOrPromise) {
+        if (writable.destroyed) return;
+        if (!writable.write(chunk)) await events_once(writable, 'drain');
+        if (writable.destroyed) return;
+      }
+      writable.end();
+    } catch (err) {
+      finish(err, null, false);
+    }
+  }
+
+  function wrap(stream, reading, writing) {
+    destroys.push(destroyer(stream, reading, writing, (err) => {
+      finish(err, null, !reading);
+    }));
+  }
+
+  function makeAsyncIterable(iterableOrPromise, final) {
+    return async function*() {
+      try {
+        for await (const chunk of await iterableOrPromise) {
+          yield chunk;
+        }
+        if (final) finish(null, null, true);
+      } catch (err) {
+        if (final) finish(err, null, true);
+        throw err;
+      }
+    }();
+  }
```

Comment:
would you mind to avoid allocating all those closures in this function? Can you make them generic and make them top level?

Reply:
This is not resolved